File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.js
function solution(topping) {
    let answer = 0;
    
    const oldBro = new Set();
    const youngBro = new Set();
    
    const cnt = new Array(10001).fill(0);
    
    topping.forEach((t) => {
        youngBro.add(t);
        cnt[t]++;
    });
    
    topping.forEach((t) => {
        if(cnt[t] > 0)
            cnt[t]--;
        if(cnt[t] == 0)
            youngBro.delete(t);

        oldBro.add(t);

        if(oldBro.size === youngBro.size)
            answer++;
    });
    
    return answer;
}

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

int solution(vector<int> topping) {
    int answer = 0;
    int cnt = 0;
    map<int, int> oldBro;
    map<int, int> youngBro;
    
    for(auto t : topping)
        youngBro[t]++;
    
    for(int i=0; i<topping.size()-1; i++){
        oldBro[topping[i]]++;
        youngBro[topping[i]]--;
        
        if(youngBro[topping[i]] == 0)
            youngBro.erase(topping[i]);
        
        if(oldBro.size() == youngBro.size())
            answer++;
    }
    
    return answer;
}